### PR TITLE
Optional inclusion of additional columns to csv at keras.callbacks.CSVLogger

### PR DESCRIPTION
**Pull request by tx1985**
### Summary
I use this PR for my own models: Enable optional inclusion of additional data columns in the csv file when using the callback CSVLogger. For example, a researcher might want to add the seconds per epoch, an identifier (same for every epoch) and the learning rate of each epoch (especially useful when using a learning rate scheduler). In my pull request, this functionality can be accessed via the parameter `additional_columns` when initialising a CSVLogger:

```
    import numpy as np
    from keras import backend as K

    def get_learning_rate(epoch=None, csvlogger=None):
        return np.round(float(K.get_value(csvlogger.model.optimizer.lr)), 5)

    csv_logger = CSVLogger('training.log',
        additional_columns={'seconds': None, 'learning_rate': get_learning_rate})
    model.fit(X_train, Y_train, callbacks=[csv_logger])
```

Parameter `additional_columns`: dictionary with key-value pairs, such that key is the name of the column to add to the csv file. If key is 'seconds' then the value is ignored and the number of seconds per epoch will be printed in the respective column 'seconds'. For other keys the value has to be a callable which must have the following two keyword arguments: `epoch` (type integer) and `model` (type Model).

### Related Issues
Related to Pull Request #4157: the current pull request includes the feature "ability to add static values to output file"

### PR Overview

- [x] This PR requires new unit tests [y/n] (make sure tests are included)
- [x] This PR requires to update the documentation [y/n] (make sure the docs are up-to-date)
- [x] This PR is backwards compatible [y/n]
- [x] This PR changes the current API [y/n] (all API changes need to be approved by fchollet)

### Further details
- I have included unit tests at `keras/tests/keras/test_callbacks.py`, which are all passing.
- I have included documentation and code examples at `keras/keras/callbacks.py`
- All code changes that I did are in the above two py-files.
- This Pull Request takes my [API Design Review suggestion](https://docs.google.com/document/d/1oU-x3l7wcZYmuk1bopnxMBs1YMyfSjQvX15k4Yn5hdA/edit#heading=h.5ojovhmssrde) a little bit further, and is a bit more clear, I believe. The latter suggestion was posted in the respective [GoogleGroup](https://groups.google.com/forum/#!topic/keras-users/lFARnrJdEcs) and had just 1 positive comment, but no further discussion.

Looking forward to feedback and improvement suggestions.

### CODE REVIEW

**Comment by fchollet** [comment]
Thanks for the PR.

This change is only relevant to few users (maybe only your own use case). Also this change is a pure addition to the callback, that would be very easy to roll out as a subclass of the callback, e.g.

```python
class CustomCSVLogger(CSVLogger):

    def __init__(self, additional_columns, **kwargs):
        super(CustomCSVLogger, self).__init__(**kwargs)
        self.epoch_start = None
        self.additional_columns = additional_columns
        for key in self.additional_columns:
            if key != 'seconds':
                argspec = inspect.getargspec if six.PY2 else inspect.getfullargspec
                args = set(argspec(self.additional_columns[key]).args)
                assert 'epoch' in args, "keyword argument 'epoch' is missing"
                assert 'model' in args, "keyword argument 'model' is missing"

    def on_epoch_begin(self, epoch, logs=None):
        if 'seconds' in self.additional_columns:
            self.epoch_start = time.time()

    def on_epoch_end(self, epoch, logs=None):
        logs = logs or {}
        for key in self.additional_columns:
            if key == 'seconds':
                now = time.time()
                logs[key] = np.round(now - self.epoch_start, 1)
            else:
                func = self.additional_columns[key]
                logs[key] = str(func(epoch=epoch, model=self.model))
        super(CustomCSVLogger, self).on_epoch_end(epoch, logs=logs)
```

I suggest that you extend the API via subclassing rather than pushing this feature upstream.

**Comment by tx1985** [comment]
Thanks @fchollet for the feedback. I agree that having a subclass is better. I added a PR #12342 , which makes this adjustment.
